File: SecRuleGen.pl
#!/usr/bin/perl
# Build mod security rules from swagger's specifications to protect a web service behind an hardened proxy
# for a better understanding of mod security see https://www.feistyduck.com/books/modsecurity-handbook and
# http://eu.wiley.com/WileyCDA/WileyTitle/productCd-1118362187.html
use strict;
use warnings;
use English;
use LWP::UserAgent;
use JSON;
no warnings "experimental::autoderef";

sub onError() {
        print "E : " . shift . "\n";
        print "Usage : ./SecGenRule.pl \"url to swagger specifications\" [whitelist] [rate limit]\n\n";
        print "\t - whitelist : list of ip adresses/subnets to exclude from rate limit rules\n";
        print "\t - rate limit : maximum number of request per user per endpoint per minute\n";
        exit 1;
}

if ( @ARGV < 1 ) {
        &onError("url not supplied");
}

my $url = $ARGV[0];
my $specs = '';
# maximum number of call to an endpoint per minute
my $maximumRateLimit = 50;
# list of IP adresses/subnets to exclude from rate limit rules
my $whitelistRateLimit = "127.0.0.1";
# Modsec id rule, start at 200, valid in most standard configuration
my $secRuleID = 200;

if ( exists($ARGV[1]) ) {
        if ( $ARGV[1] =~ m/[0-9\.,\/]/ ) {
                $whitelistRateLimit = $ARGV[1];
        }
        else {
                &onError("whitelist invalid");
        }
}

if ( exists($ARGV[2]) ) {
        if ( $ARGV[2] =~ m/[0-9]/ ) {
                $maximumRateLimit = $ARGV[2];
        }
        else {
                &onError("rate limit invalid");
        }
}

# Downlad the json given as argument
sub DownlSpecs() {
        my ($dest_file) = @_;
        if ( $dest_file =~ m/^http/ ) {
                $dest_file =~ s/^.*\/\///;
        }
        $dest_file =~ s/\/.*$//;
        $dest_file .= ".json";
        my $ua = LWP::UserAgent->new();
        my $req = HTTP::Request->new(GET => $url);
        my $res = $ua->request($req);
        if ( $res->is_success ) {
                if ( -e $dest_file ) {
                        unlink $dest_file;
                }
                open (TARGET, '>', $dest_file) or die $!;
                print TARGET $res->content;
                close(TARGET);
        }
        else {
                &onError("invalid url");
        }
        return $dest_file;
}

# Generate the ID for every modsec rule
sub SecRuleIDGen() {
        $secRuleID++;
        return $secRuleID;
}

# Dig into swagger argument definition to extract every possible argument
sub ManageComplexArg {
        my ($root, $ref) = @_;
        my $parameters = '';
        $ref  =~ s/^.*\///g;
        my @properties = keys $specs->{"definitions"}{$ref}{"properties"};
        foreach my $property (sort @properties) {
                my $validValues = '';
                my $newParam = '';
                my $parsedArgs = '';
                my $isComplex = 0;
                if ( exists($specs->{"definitions"}{$ref}{"properties"}{$property}{"items"}{"\$ref"}) ) {
                        $isComplex = 1;
                        $parameters .= $property . "\.*|";
                }
                if ( exists($specs->{"definitions"}{$ref}{"properties"}{$property}{"\$ref"}) ) {
                        $isComplex = 1;
                        $parameters .= $property . "\.*|";
                }
                if ( $specs->{"definitions"}{$ref}{"properties"}{$property}{"type"} =~ m/object/ ) {
                        $isComplex = 1;
                        $parameters .= $property . "\.*|";
                }
                if ( !$isComplex ) {
                        if ( $root =~ m/NULL/ ) {
                                $newParam = $property;
                        }
                        else {
                                $newParam = $root . "\\." . $property;
                        }
                        if ( exists($specs->{"definitions"}{$ref}{"properties"}{$property}{"enum"}) && index($parsedArgs, $property . ",") == -1 && index($parsedArgs, "," . $property) == -1 ) {
                                $parsedArgs .= $property . ",";
                                my @values = keys $specs->{"definitions"}{$ref}{"properties"}{$property}{"enum"};
                                foreach my $value (@values) {
                                        $validValues .= $specs->{"definitions"}{$ref}{"properties"}{$property}{"enum"}[$value] . "|";
                                }
                        }
                        if ( $parameters !~ "|" . $newParam . "|" ) {
                                $parameters .= $newParam . "|";
                        }
                }
        }
        return $parameters;
}

# Remove var from url (such as {id} or {email}), replace it with a regex matching anything exept "/"
sub CleanArgFromUrl {
        my ($url) = @_;
        $url =~ s/\{[a-zA-Z0-9\-_\.]*\}/\(\?\!\(\.\*\/\)\)/;
        if ( $url =~ m/\{.*\}/) {
                &CleanArgFromUrl($url);
        }
        else {
                return($url);
        }
}

# Generate a valid LocationMatch directive for apache from the given URL
sub BuildModsecLocationMatch {
        my ($url) = @_;
        if ( $url =~ m/\{.*\}/ ) {
                $url = &CleanArgFromUrl($url);
        }
        $url = "<LocationMatch \"^" . $url . "\$\">";
        return($url);
}

# Generate a modsec's arg_names rule from given list of argument, basicaly we are checking if the  supplied argument is listed in the spec
sub BuildModsecArgNamesRule {
        my ($arg) = @_;
        $arg =~ s/\|$//;
        my $rule = "SecRule ARGS_NAMES \"!^(?:" . $arg . ")\$\" \"phase:2,t:none,deny,id:'" . &SecRuleIDGen . "',status:400,msg:'Unrecognized argument',logdata:%{MATCHED_VAR},setenv:ARGNAMEERROR=%{MATCHED_VAR}\"";
        return $rule;
}

# Generate a modsec's request_method rule from given list of methods extracted from swagger's specifications
sub BuildModsecReqMethodRule {
        my @methods = @_;
        my $rule = "SecRule REQUEST_METHOD \"!^(?:";
        foreach my $method (@methods) {
                $rule .= uc($method) . "|";
        }
        $rule =~ s/\|$/\|OPTIONS/;
        $rule .= ")\$\" \"phase:2,t:none,deny,id:'" . &SecRuleIDGen . "',status:405,msg:'Unauthorized method',logdata:%{REQUEST_METHOD},setenv:METHODERROR\"";
        return $rule;
}

# Build a modsec request header rule to check if the request content-type match the on specified in swagger's endpoint description
sub BuildModsecContentTypeRule {
        my ($contentType) = @_;
        $contentType =~ s/\|$//;
        my $rule = "SecRule REQUEST_HEADERS:Content-Type \"!\@rx (?i)^(";
        $rule .= $contentType;
        $rule .= ")\" \"phase:1,t:none,deny,id:'" . &SecRuleIDGen . "',status:400,msg:'Invalid content-type',logdata:%{MATCHED_VAR},setenv:CTYPEERROR\"";
        return $rule;
}

# Build a rule to check if a post parameter's content is valid against swagger's supplied format
sub BuildModsecDataTypeRule() {
        my ($argName, $dataType, $dataFormat) = @_;
        my $matchedFormat = 0;
        my $rule = "SecRule ARGS:" . $argName . " \"!^(?:";
        if ( $dataType =~ m/integer/) {
                $rule .= "[0-9\-]*";
                $matchedFormat = 1;
        }
        if ( $dataType =~ m/number/) {
                $rule .= "[0-9\.,\-]*";
                $matchedFormat = 1;
        }
        if ( $dataType =~ m/boolean/) {
                $rule .= "0|1|true|false|TRUE|FALSE";
                $matchedFormat = 1;
        }
        if ( $matchedFormat ) {
                $rule .= ")\$\" \"phase:2,t:none,deny,id:'" . &SecRuleIDGen . "',status:400,msg:'Bad content format',logdata:%{MATCHED_VAR},setenv:CONTENTERROR=%{MATCHED_VAR}\"";
                return $rule;
        }
        else {
                &onError("data format unknown for arg $argName");
        }
}

# Build a batch of rule to control the max number request per user per minute
sub BuildModSecRateLimitRule() {
        my ($endpoint) = @_;
        $endpoint =~ s/\{.*\}//;
        my $id = &SecRuleIDGen;
        my $rule = "SecRule REMOTE_ADDR \"\@ipMatch " . $whitelistRateLimit . "\" \"skipAfter:IGNORE_RATELIMIT_" . $id . ",nolog,id:'" . &SecRuleIDGen . "'\"\n\t\t";
        $rule .= "SecAction \"initcol:ip=\%{REMOTE_ADDR}_\%{REQUEST_HEADERS.User-Agent},pass,nolog,id:'" . &SecRuleIDGen . "'\"\n\t\t";
        $rule .= "SecAction \"phase:5,deprecatevar:ip." . $endpoint . "=" . $maximumRateLimit . "/60,pass,nolog,id:'" . &SecRuleIDGen . "'\"\n\t\t";
        $rule .= "SecAction \"phase:2,pass,setvar:ip." . $endpoint . "=+1,nolog,id:'" . &SecRuleIDGen . "'\"\n\t\t";
        $rule .= "SecRule IP:" . $endpoint . " \"\@gt " . $maximumRateLimit . "\" \"phase:2,pause:300,deny,setenv:RATELIMITED,skip:1,id:'" . &SecRuleIDGen . "',status:400,msg:'too many request per minute',logdata:%{MATCHED_VAR}\"\n\t\t";
        $rule .= "SecMarker IGNORE_RATELIMIT_" . $id;
        return $rule;
}

# Main routine, extract infos from the downloaded json and call subroutine to build modsec rules
sub ParseJson {
        my ($file) = @_;
        open ('JSON', '<', $file) or &onError("can't open file");
        my $content;
        while ( <JSON> ){
                $content .= $_;
        }
        $specs = decode_json($content);
        my @paths = keys $specs->{"paths"};
        foreach my $url (@paths) {
                my @methods = keys $specs->{"paths"}{$url};
                if ( defined($methods[0]) ) {
                        print "\t" . &BuildModsecLocationMatch($url) . "\n";
                        print "\t\t" . &BuildModsecReqMethodRule(@methods) . "\n";
                        print "\t\t" . &BuildModSecRateLimitRule($url) . "\n";
                        my $paramList = '';
                        my $contentTypeList = '';
                        my $parsedArgs = '';
                        foreach my $method (@methods) {
                                if ( exists($specs->{"paths"}{$url}{$method}{"parameters"}) ) {
                                        my @parameters = keys $specs->{"paths"}{$url}{$method}{"parameters"};
                                        foreach my $parameter (@parameters) {
                                                my $validValues = '';
                                                # If the argument has a definition
                                                if ( exists($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"schema"}{"\$ref"}) ) {
                                                        # And we haven't parsed it before
                                                        if ( index($paramList, &ManageComplexArg("NULL", $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"schema"}{"\$ref"})) == -1 ) {
                                                                $paramList .= &ManageComplexArg("NULL", $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"schema"}{"\$ref"});
                                                        }
                                                }
                                                else {
                                                        if ( index($paramList, "|" . $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"} . "|") == -1 ) {
                                                                $paramList .= $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"} . "|";
                                                        }
                                                }
                                                # If variable content are enumerated we build the matching rule
                                                if ( exists($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"enum"}) && index($parsedArgs, $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"} . "," ) == -1 && index($parsedArgs, "," . $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"}) == -1 ) {
                                                        $parsedArgs .= $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"} . ",";
                                                        my @values = keys $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"enum"};
                                                        foreach my $value (@values) {
                                                                $validValues .= $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"enum"}[$value] . "|";
                                                        }
                                                }
                                                # If variable data type is supplied we build the matching rule
                                                if ( exists($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"type"}) ) {
                                                        # Skipping if data type is "string" or if data format is not supplied
                                                        if ( ($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"type"} !~ m/string/) && (exists($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"format"})) ) {
                                                                print "\t\t" . &BuildModsecDataTypeRule($specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"name"}, $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"type"}, $specs->{"paths"}{$url}{$method}{"parameters"}[$parameter]{"format"}) . "\n";
                                                        }
                                                }
                                        }
                                }
                                # If a content type for input is specified we build the matching rule
                                if ( exists($specs->{"paths"}{$url}{$method}{"consumes"}) ) {
                                        my @contentTypes =  keys $specs->{"paths"}{$url}{$method}{"consumes"};
                                        foreach my $contentType (@contentTypes){
                                                $contentTypeList .= $specs->{"paths"}{$url}{$method}{"consumes"}[$contentType] . "|";
                                        }
                                }
                        }
                        if ( length($paramList) > 0 ) {
                                print "\t\t" . &BuildModsecArgNamesRule($paramList) . "\n";
                        }
                        if ( length($contentTypeList) > 0 ) {
                                print "\t\t" . &BuildModsecContentTypeRule($contentTypeList) . "\n";
                        }
                        print "\t</LocationMatch>\n";
                }
        }
        print "\tHeader always set Retry-After \"60\" env=RATELIMITED\n";
        print "\tHeader always set izigloo-error-id \"PROXY-000\" env=RATELIMITED\n";
        print "\tHeader always set smartdiag-error-message \"too many request in the last minute\" env=RATELIMITED\n";
        print "\tHeader always set izigloo-error-id \"PROXY-001\" env=ARGNAMEERROR\n";
        print "\tHeader always set smartdiag-error-message \"wrong argument supplied : %{ARGNAMEERROR}e\" env=ARGNAMEERROR\n";
        print "\tHeader always set izigloo-error-id \"PROXY-002\" env=CTYPERROR\n";
        print "\tHeader always set smartdiag-error-message \"wrong content type\" env=CTYPERROR\n";
        print "\tHeader always set izigloo-error-id \"PROXY-003\" env=METHODERROR\n";
        print "\tHeader always set smartdiag-error-message \"wrong method used\" env=METHODERROR\n";
        print "\tHeader always set izigloo-error-id \"PROXY-004\" env=CONTENTERROR\n";
        print "\tHeader always set smartdiag-error-message \"wrong content in one argument : %{CONTENTERROR}e\" env=CONTENTERROR\n";
}

my $file = &DownlSpecs($url);
&ParseJson($file);
unlink($file);

File: SecRuleGen.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"io/ioutil"
	"encoding/json"
    "strings"
    "strconv"
    "sort"
    //"bufio"
	"regexp"
)

type Config struct {
    Url string `json:"url"`
    Ratelimit int `json:"ratelimit"`
    RatelimitWhitelist string `json:"ratelimit_whitelist"`
    RestrictedEndpoints []RestrictedEndpoint `json:"restricted_endpoints"`
    WebServer string `json:"webserver"`
    Verbose bool `json:"verbose"`
}
type RestrictedEndpoint struct {
    Path string `json:"path"`
    IpAllowed string `json:"ip_allowed"`
}

type Endpoint struct {
    Url string
    Methods []Method
}
type Method struct {
	Name string
	Parameters []EndpointParameter
}
type EndpointParameter struct {
    Name string `json:"name"`
    In string `json:"in"`
    Description string `json:"description"`
    Format string `json:"format"`
    Type string `json:"type"`
    Required bool `json:"required"`
    Enum []string `json:"enum"`
    Schema ParameterSchema `json:"schema"`
}
type ParameterSchema struct {
	Ref string `json:"$ref"`
	Type string `json:"type"`
}

type Definition struct {
	Name string
	Type string `json:"type"`
	Properties []PropertyDefinition `json:"properties"`
}
type PropertyDefinition struct {
	Name string
	Type string `json:"type"`
	Format string `json:"format"`
	Ref string `json:"$ref"`
	Items string `json:"items"`
	Enum []string `json:"enum"`
}

var modsecRuleID = 30000
func _GetModsecRuleID() int {
	modsecRuleID = modsecRuleID + 1
	return modsecRuleID
}

func ReadConfigFile(s string) Config {
    configFile, err := os.Open(s)
    defer configFile.Close()
    if err != nil {
        fmt.Fprintf(os.Stderr, "Err : unable to read %s\n", s)
        os.Exit(1)
    }
    bytes, _ := ioutil.ReadAll(configFile)
    var config Config
    json.Unmarshal(bytes, &config)
    return config
}

func GetSwaggerSpec(s string) interface{} {
    response, err := http.Get(s)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Err : unable to download %s\n", s)
        os.Exit(1)
    }
    body, err := ioutil.ReadAll(response.Body)
    var swaggerSpec interface{}
    json.Unmarshal(body, &swaggerSpec)
    return swaggerSpec
}

func GetEndpointList(s map[string]interface{}) []Endpoint {
    var list []Endpoint
    for endpoint, availableMethods := range s["paths"].(map[string]interface{}) {
    	var methods []Method
    	var parameters []EndpointParameter
    	for method, description := range availableMethods.(map[string]interface{}){
    		for property, v := range description.(map[string]interface{}){
    			if property == "parameters" {
    				switch x := v.(type) {
						case []interface{}:
    						for _, e := range x {
        						var p EndpointParameter
        						s, _ := json.Marshal(e)
        						json.Unmarshal(s, &p)
        						if len(p.Schema.Ref) > 0 {
        							for _, v := range _ParseParameterDefinition(p.Name, p.Schema.Ref) {
        								parameters = append(parameters, v)
        							}
        						} else {
        							p.Type = "string"
        							parameters = append(parameters, p)
        						}
    						}
						default:
    						fmt.Printf("can't parse parameter %T\n", v)
					}
    			}
    		}
    		methods = append(methods, Method{method, parameters})
    	}
    	list = append(list, Endpoint{endpoint, methods})
    }
	testSlash := regexp.MustCompile("/")
    sort.Slice(list, func(i, j int) bool {
		matchesI := testSlash.FindAllStringIndex(list[i].Url, -1)
		matchesJ := testSlash.FindAllStringIndex(list[j].Url, -1)
    	return len(matchesI) < len(matchesJ)
    	})
	return list
}

func GetDefinitionList(s map[string]interface{}) []Definition {
	var definitions []Definition
	for name, properties := range s["definitions"].(map[string]interface{}) {
		var d Definition
		m := properties.(map[string]interface{})
		d.Name = "#/definitions/" + name
		d.Type = m["type"].(string)
		fmt.Println("Definition : ", d.Name)
		if d.Type == "object" {
			var p PropertyDefinition
			for name, property := range m["properties"].(map[string]interface{}) {
				p.Name = name
				fmt.Println("\tproperty : ", name, " : ", property)
				for k, v := range property.(map[string]interface{}) {
					switch k {
					case "type":
						p.Type = v.(string)
					case "format":
						p.Format = v.(string)
					//case "items":
					//	p.Items = v.(string)
					case "$ref":
						p.Ref = v.(string)
					case "enum":
						for _, val := range v.([]interface{}){
							p.Enum = append(p.Enum, val.(string))
						}
					}
				}
				fmt.Println("\t-> result : ", p)
			}
			//d.Properties = append(d.Properties, p)
		}
		definitions = append(definitions, d)
	}
	return definitions
}

func GenerateRules(e []Endpoint, c Config ) []string {
    var rules []string
    var outputStyle string
    //var ratelimit string = c.Ratelimit
    if c.WebServer == "apache" || c.WebServer == "nginx" {
        outputStyle = c.WebServer
    } else {
        return rules
    }
    for _, endpoint := range e {
    	rules = append(rules, _GenerateLocationBlockHeader(endpoint, outputStyle))
    	if len(_GenerateSourceIpAddrRules(endpoint, c)) > 0 {
    		for _, r := range _GenerateSourceIpAddrRules(endpoint, c) {
    			rules = append(rules, r)
    		}
    	}
    	rules = append(rules, _GenerateMethodRule(endpoint))
    	if len(_GenerateRatelimitRules(c)) > 0{
    		for _, line := range _GenerateRatelimitRules(c) {
    			rules = append(rules, line)
    		}
    	}
    	for _, line := range _GeneratePerMethodRules(endpoint){
    		rules = append(rules, line)
    	}
    	rules = append(rules, _GenerateLocationBlockFooter(outputStyle))
    }
    return rules
}

func _ParseParameterDefinition(n string, r string) []EndpointParameter {
	var parameters []EndpointParameter

	return parameters
}

func _GeneratePerMethodRules(e Endpoint) []string {
	var rules []string
	for _, method := range e.Methods {
		id := strconv.Itoa(_GetModsecRuleID())
		rules = append(rules, "SecRule REQUEST_METHOD \"!^(?:"+ strings.ToUpper(method.Name) + ")$\" \"skipAfter:FILTER_BY_METHOD_" + id + ",nolog,id:'" + id + "'\"")

		rules = append(rules, "SecMarker FILTER_BY_METHOD_" + id)
	}
	return rules
}

func _GenerateRatelimitRules(c Config) []string {
	var rules []string
	if len(c.RatelimitWhitelist) == 0 || c.Ratelimit == 0 {
		return rules
	}
	id := strconv.Itoa(_GetModsecRuleID())
	rules = append(rules, "SecRule REMOTE_ADDR \"@ipMatch " + c.RatelimitWhitelist + "\" \"skipAfter:IGNORE_RATELIMIT_" + id + ",nolog,id:'" + id + "'\"")
	rules = append(rules, "SecAction \"initcol:ip=%{REMOTE_ADDR}_%{REQUEST_HEADERS.User-Agent},pass,nolog,id:'" + strconv.Itoa(_GetModsecRuleID()) + "'\"")
	rules = append(rules, "SecAction \"phase:5,deprecatevar:ip./v1/professionnels/chantiers/count/maj-crm=100/60,pass,nolog,id:'" + strconv.Itoa(_GetModsecRuleID()) + "'\"")
	rules = append(rules, "SecAction \"phase:2,pass,setvar:ip./v1/professionnels/chantiers/count/maj-crm=+1,nolog,id:'" + strconv.Itoa(_GetModsecRuleID()) + "'\"")
	rules = append(rules, "SecRule IP:/v1/professionnels/chantiers/count/maj-crm \"@gt " + strconv.Itoa(c.Ratelimit) + "\" \"phase:2,pause:300,deny,setenv:RATELIMITED,skip:1,id:'" + strconv.Itoa(_GetModsecRuleID()) + "',status:400,msg:'too many request per minute',logdata:%{MATCHED_VAR}\"")
	rules = append(rules, "SecMarker IGNORE_RATELIMIT_" + id)
	return rules
}

func _GenerateMethodRule(e Endpoint) string{
	rule := "SecRule REQUEST_METHOD \"!^(?:"
	var methods string
	for _ , method := range e.Methods {
		if len(methods) == 0 {
			methods = methods + strings.ToUpper(method.Name)
		} else {
			methods = methods + "|" + strings.ToUpper(method.Name)
		}
	}
	methods = methods + "|OPTIONS"
	rule = rule + methods
	rule = rule + ")$\" \"phase:2,t:none,deny,id:'" + strconv.Itoa(_GetModsecRuleID()) + "',status:405,msg:'Unauthorize method',logdata:%{REQUEST_METHOD},setenv:METHODERROR\""
	return rule
}

func _GenerateSourceIpAddrRules(e Endpoint, c Config ) []string {
	var rules []string
	if len(c.RestrictedEndpoints) > 0{
		for _, v := range c.RestrictedEndpoints {
			if v.Path == e.Url {
				id1 := strconv.Itoa(_GetModsecRuleID())
				rules = append(rules, "SecRule REMOTE_ADDR \"@ipMatch " + v.IpAllowed + "\" \"id:'" + id1 + "',skipAfter:IP_IS_ALLOWED_" + id1 + ",nolog\"")
				rules = append(rules, "SecAction \"deny,id:'" + strconv.Itoa(_GetModsecRuleID()) + "',log,msg:'IP not allowed on this endpoint',logdata:%{MATCHED_VAR}\"")
				rules = append(rules, "SecMarker IP_IS_ALLOWED_" + id1)
			}
		}
	}
	return rules
}

func _GenerateLocationBlockHeader(e Endpoint, s string) string{
    var locationBlockHeader string
    url := _URLParameterToRegex(e)
    if s == "apache" {
        locationBlockHeader = "<LocationMatch \"^"+ url + "$\">"
    }
    if s == "nginx" {
    	locationBlockHeader = "location " + url + "{ modsecurity_rules '"
    }
    return locationBlockHeader
}

func _GenerateLocationBlockFooter(s string) string {
	var block string
	if s == "apache" {
		block = "</LocationMatch>"
	}
	if s == "nginx" {
		block =   "';}"
	}
	return block
}

func _URLParameterToRegex(e Endpoint) string {
	buff := e
	if strings.ContainsAny("{", buff.Url) == false {
		return buff.Url
	}
	urlParameter := buff.Url[(strings.Index(buff.Url, "{")):strings.Index(buff.Url, "}")+1]
    for _, method := range buff.Methods {
        for _, parameter := range method.Parameters {
            if "{" + parameter.Name + "}" == urlParameter {
				idxS := strings.Index(buff.Url, urlParameter)
				idxE := idxS + len(urlParameter)
            	buff.Url = e.Url[:idxS] + "(" + _TypeToRegex(parameter.Type) + ")" + e.Url[idxE:]
            }
        }
		if strings.ContainsAny("{", buff.Url) == false {
			break
		}
    }
	if strings.ContainsAny("{", buff.Url) == true {
		return _URLParameterToRegex(buff)
	}
    return buff.Url
}

func _TypeToRegex(t string) string{
    switch t {
    case "boolean":
        return "[0-1]"
    case "integer":
        return "[+-]?[0-9]*"
    case "number":
        return "[+-]?[0-9\\.,]*"
    case "string":
        return "[^/]*"
    default:
        return ""
    }
}

func ReplaceDefinitions (e []Endpoint, d []Definition) []Endpoint {
	return e
}

func main() {
    if len(os.Args) != 2 {
        fmt.Fprintf(os.Stderr, "Usage: %s config.json\n", os.Args[0])
        os.Exit(1)
    }
    config := ReadConfigFile(os.Args[1])
    swaggerSpecs := GetSwaggerSpec(config.Url).(map[string]interface{})
    fmt.Println("Version of swagger specifications : ", swaggerSpecs["swagger"], "\nParsing its content...")
	endpoints := ReplaceDefinitions(GetEndpointList(swaggerSpecs), GetDefinitionList(swaggerSpecs))
	GenerateRules(endpoints, config)
    //rules := GenerateRules(endpoints, config)
    //for _, v := range rules {
    //	fmt.Println(v)
    //}
    //fmt.Println(definitions)
}